internal/datastore/common/revisions/optimized.go

```diff
@@ -43,45 +41,71 @@ func (cor *CachedOptimizedRevisions) OptimizedRevision(ctx context.Context) (dat
 	defer span.End()
 
 	localNow := cor.clockFn.Now()
-	lastRevision := cor.lastQuantizedRevision.get()
-	if localNow.Before(lastRevision.validThrough) {
-		log.Ctx(ctx).Debug().Time("now", localNow).Time("valid", lastRevision.validThrough).Msg("returning cached revision")
-		span.AddEvent("returning cached revision")
-		return lastRevision.revision, nil
+
+	cor.Lock()
+
+	// Prune the candidates that have definitely expired
+	var numToDrop uint
+	for _, candidate := range cor.candidates {
+		if candidate.validThrough.Add(cor.maxRevisionStaleness).Before(localNow) {
+			numToDrop++
+		} else {
+			break
+		}
+	}
+	cor.candidates = cor.candidates[numToDrop:]
+
+	// Subtract a random amount of time from now, to let barely expired candidates get selected
+	adjustedNow := localNow
+	if cor.maxRevisionStaleness > 0 {
+		adjustedNow = localNow.Add(-1 * time.Duration(rand.Int63n(cor.maxRevisionStaleness.Nanoseconds())) * time.Nanosecond)
+	}
+
+	for _, candidate := range cor.candidates {
+		if candidate.validThrough.After(adjustedNow) {
+			log.Ctx(ctx).Debug().Time("now", localNow).Time("valid", candidate.validThrough).Msg("returning cached revision")
+			span.AddEvent("returning cached revision")
+			cor.Unlock()
+			return candidate.revision, nil
+		}
 	}
 
-	lastQuantizedRevision, err, _ := cor.updateGroup.Do("", func() (interface{}, error) {
-		log.Ctx(ctx).Debug().Time("now", localNow).Time("valid", lastRevision.validThrough).Msg("computing new revision")
+	cor.Unlock()
+
+	newQuantizedRevision, err, _ := cor.updateGroup.Do("", func() (interface{}, error) {
+		log.Ctx(ctx).Debug().Time("now", localNow).Msg("computing new revision")
 		span.AddEvent("computing new revision")
 
 		optimized, validFor, err := cor.optimizedFunc(ctx)
 		if err != nil {
 			return nil, fmt.Errorf("unable to compute optimized revision: %w", err)
 		}
 
-		rvt := localNow.
-			Add(validFor).
-			Add(cor.maxRevisionStaleness)
-		cor.lastQuantizedRevision.set(validRevision{optimized, rvt})
+		rvt := localNow.Add(validFor)
+		cor.Lock()
+		defer cor.Unlock()
+		cor.candidates = append(cor.candidates, validRevision{optimized, rvt})
 		log.Ctx(ctx).Debug().Time("now", localNow).Time("valid", rvt).Stringer("validFor", validFor).Msg("setting valid through")
 
 		return optimized, nil
 	})
 	if err != nil {
 		return datastore.NoRevision, err
 	}
-	return lastQuantizedRevision.(datastore.Revision), err
+	return newQuantizedRevision.(datastore.Revision), err
 }
 
 // CachedOptimizedRevisions does caching and deduplication for requests for optimized revisions.
 type CachedOptimizedRevisions struct {
+	sync.Mutex
```

Comment:
I came across https://pkg.go.dev/github.com/puzpuzpuz/xsync/v2#RBMutex recently which based on the docs + benchmarks should have less contention for readers. We could also write our own benchmarks for optimized revision to see if it helps. 

Maybe this is premature, but every request goes through this path.

Reply:
In actual tests this seemed to make it slower.